### PR TITLE
Dataloader Upgrades

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
1. Before, the `dataloder_spec`, `batch_size`, and `dataloader_hparams` were passed as areguments into the trainer. Now, the trainer is initialized with a dataloader (or a dataloader, split_fn, and preprocessing_fn tuple). This change makes the `DataloaderSpec` optional and hidden to the user for simple datasets that do not require custom preprocessing or split functions.

2. Removed `dataloader_to_device` and replaced it with explicit calls in the training loop to 1) move data onto the device, and 2) execute the preprocessing fn. The preprocessing fn is renamed to device transformation fn. Removed the option to execute the device transformation fn in a cuda stream, since that did not add any performance improvement. When using memory pinning, the `batch_to_device` should be a no-op, since the dataloader would have already moved the data onto the GPU.

TODO:
- [x] Regression test on resnet base to ensure no throughput or accuracy degredations

### CODE REVIEW

File: composer/models/gpt2/scaling_laws_generator.py [hunk 1]
```diff
@@ -177,10 +177,7 @@ def parse_args():
     'eval_batch_size': 8,
     'seed': 17,
     'accelerator': {
-        'gpu': {
-            'n_gpus': 1,
```
Comment (jbloxham) [inline composer/models/gpt2/scaling_laws_generator.py:181]: Woops, looks like I missed this. Thanks for catching!

**Review by jbloxham — COMMENTED** [review]
This looks good - I'm very happy to see `DataloaderSpec` losing favor. The only thing I wonder is whether `split_fn` and `device_transform_fn` could be removed. The former I think is unnecessary if we just load `N` microbatches instead of 1 batch, and the latter could be replaced with explicit augmentations?

File: composer/datasets/dataloader.py [hunk 2]
```diff
@@ -79,47 +78,69 @@ def __next__(self) -> Batch:
...
-            batch_size=batch_size,
-            shuffle=False,  # set in the sampler
-            num_workers=self.num_workers,
-            pin_memory=self.pin_memory,
-            drop_last=dataloader_spec.drop_last,
-            sampler=sampler,
-            collate_fn=dataloader_spec.collate_fn,
-            worker_init_fn=dataloader_spec.worker_init_fn,
-            multiprocessing_context=dataloader_spec.multiprocessing_context,
-            generator=dataloader_spec.generator,
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/datasets/dataloader.py:121]: > This looks good - I'm very happy to see `DataloaderSpec` losing favor. The only thing I wonder is whether `split_fn` and `device_transform_fn` could be removed. The former I think is unnecessary if we just load `N` microbatches instead of 1 batch, and the latter could be replaced with explicit augmentations?

Re: split_fn, I think we want to have the dataloader yield entire batches instead of microbatches, but @abhi-mosaic can speak more to this one.

Re: device_transform_fn: The transforms that happen in the dataset (e.g. resnet crop/resize) happen on the cpu. This allows for GPU transformations that are specified by the dataset (like what imagenet does with the batchnorm).

File: composer/models/gpt2/scaling_laws_generator.py (hunk 1 above)
Comment (ravi-mosaicml) [reply composer/models/gpt2/scaling_laws_generator.py:181]: (This code doesn't have test cases :D )

File: composer/datasets/dataloader.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/datasets/dataloader.py:121]: Multiprocessing context is undocumented so I presume it's deprecated? https://pytorch.org/docs/stable/data.html

generator -- Since we're passing a sampler to the dataloader, the generator argument is unused. The dataset can specify a custom generator in the sampler constructor, if desired.

[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/datasets/dataloader.py:121]: I think it may be simpler to keep our system of loading full batches for now... but probably worth revisiting later. Now that we're getting into multi-node training, if we run a workload on 1 vs. 8 vs. 64 GPUs, the visible per_gpu_batch_size will be different for any algorithm like Mixup that would like to perform an operation across-batch. So should we make it possible to always alter the full batch (across GPUs), or should we expose the microbatching somehow to algorithms....

**Review by jbloxham — APPROVED** [review]

**Comment by ravi-mosaicml** [comment]
Manually tested on a gpu instance. Throughput for a 2-wide box was 1290.

**Comment by abhi-mosaic** [comment]
I'm getting a bit confused as to how the `train_batch_size` is computed / saved, just to clarify:

- `Trainer` no longer knows what the batch size is at init, instead it looks at its device's dataloader, and the world size, and computes `train_batch_size`, and then creates `State` with this value
- `TrainerHparams`, which has a field for `train_batch_size`, will carefully create each of the device dataloaders such that they have batch sizes of `train_batch_size` / `world_size`, such that when they are re-interpreted by `Trainer`, it will be computed correctly

Also, can we rename `total_batch_size` -> `train_batch_size`? It's always weirded me out

**Review by abhi-mosaic — APPROVED** [review]
Discussed with @ravi-mosaicml and I think this looks good to me pending throughput sanity checks on ImageNet.
